rust/lance-index/src/vector/ivf/shuffler.rs

```diff
@@ -95,69 +260,84 @@ pub async fn shuffle_dataset(
+                    let nan_filter = KeepFiniteVectors::new(&column);
+
+                    tokio::task::spawn(async move {
+                        let mut batch = b?;
+
+                        if !partition_map.is_empty() {
+                            let row_ids = batch.column_by_name(ROW_ID).ok_or(Error::Index {
+                                message: "column does not exist".to_string(),
+                                location: location!(),
+                            })?;
+                            let part_ids = UInt32Array::from_iter(
+                                row_ids.as_primitive::<UInt64Type>().values().iter().map(
+                                    |row_id| {
+                                        let row_addr = RowAddress::new_from_id(*row_id);
+                                        partition_map[row_addr.fragment_id() as usize]
+                                            [row_addr.row_id() as usize]
+                                    },
+                                ),
                             );
-                            assert_eq!(indices.len(), batch.num_rows() - part_ids.null_count());
-                            batch = batch.take(&indices)?;
+                            let part_ids = UInt32Array::from(part_ids);
+                            batch = batch
+                                .try_with_column(
+                                    Field::new(PART_ID_COLUMN, part_ids.data_type().clone(), true),
+                                    Arc::new(part_ids.clone()),
+                                )
+                                .expect("failed to add part id column");
+
+                            if part_ids.null_count() > 0 {
+                                info!(
+                                    "Filter out rows without valid partition IDs: null_count={}",
+                                    part_ids.null_count()
+                                );
+                                let indices = UInt32Array::from_iter(
+                                    part_ids
+                                        .iter()
+                                        .enumerate()
+                                        .filter_map(|(idx, v)| v.map(|_| idx as u32)),
+                                );
+                                assert_eq!(indices.len(), batch.num_rows() - part_ids.null_count());
+                                batch = batch.take(&indices)?;
+                            }
                         }
-                    }
 
-                    // Filter out NaNs/Infs
-                    batch = nan_filter.transform(&batch)?;
-
-                    ivf.transform(&batch)
+                        // Filter out NaNs/Infs
+                        let minibatch_size = std::env::var("LANCE_SHUFFLE_BATCH_SIZE")
+                            .unwrap_or("64".to_string())
+                            .parse::<usize>()
+                            .unwrap_or(64);
```

Comment:
If someone does specify this but there is a parsing error, I think we should either log or error, rather than silently fall back to the default.